winit: initial minibrowser

delan opened the pull request:
This patch introduces a new minibrowser mode, enabled by running the winit port with --minibrowser.

The minibrowser has a location bar and a go button. The location bar updates to reflect the URL of the top-level browsing context, unless the user has started editing the location without clicking go.

It has been tested on Linux with X11, but should also work with Wayland and on Windows and macOS.

![image](https://github.com/servo/servo/assets/465303/ccf29abc-d10c-4ff0-808b-eb15cf8624cb)

Some changes to the compositor were necessary to allow Servo and our minibrowser code to cooperatively draw to the same OpenGL framebuffer. Normally the compositor “presents” each new frame (swapping buffers) as soon as WebRender has drawn to it, but we need to make sure that both WebRender and the minibrowser are [repainted on every frame](https://gamedev.stackexchange.com/q/7320). Failing to do so leads to strange glitches where new frames get overwritten with stale framebuffer contents.

We add an external­_present mode that instead notifies the embedder (via the constellation) that a new frame is ready, allowing the embedder to draw other things to the framebuffer. In external­_present mode, the compositor does not run again until the embedder calls present, to ensure consistency of WebRender output, and IO­Compositor­::clear­_background no longer clears the whole framebuffer, only the viewport area.

Some optimisations were also necessary to make this performant:

* we add a Browser field that tracks the current top-level URL as a string, so we can avoid repeated serialisation
* we vendor and modify EguiGlow to retain its shapes, so we can repaint the minibrowser without an egui update
* when egui needs an update, we defer it through a winit redraw request, so we can coalesce updates over events

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #29930
- [ ] These changes partially involve #30049

<!-- Either: -->
- [ ] There are tests for these changes OR
- [x] These changes do not require tests because it would be impractical to test the minibrowser directly, but we may be able to unify the codepaths so the minibrowser is indirectly tested with WPT

delan committed:
winit: add minibrowser feature that depends on egui{,-winit}

delan committed:
winit: carve out some space at the top of headed windows

delan committed:
winit: minimal toolbar and egui/winit integration (but no painting)

delan committed:
winit: try to paint with egui_glow (doesn’t work yet)

delan committed:
winit: add comment about toolbar size

atbrakhi committed:
Add framebuffer object, set it as glow's target

delan committed:
compositing: clear only the viewport, not the whole framebuffer

atbrakhi committed:
plumb the actual size of the egui toolbar to webrender

atbrakhi committed:
fix formatting

delan committed:
winit: fix crash when fbo is zero

delan commented on ports/winit/app.rs:
```diff
@@ -62,6 +71,83 @@ impl App {
+        // glow needs to set framebuffer as a target
+        unsafe {
+            gl.bind_framebuffer(glow::FRAMEBUFFER, framebuffer_object);
+        };
```
when i run this now, the fbo number is zero, which apparently tells [glBindFramebuffer](https://registry.khronos.org/OpenGL-Refpages/gl4/html/glBindFramebuffer.xhtml) to break any existing binding. i wonder if we really need this?

delan replied:
removed in e9f8c3c0bbbdc6166f9006dc69d88a11103e2b2d

delan committed:
winit: don’t bother binding the framebuffer object

atbrakhi committed:
winit: remove unsafe and get toolbar_height

atbrakhi commented on ports/winit/app.rs:
```diff
@@ -95,9 +97,9 @@ impl App {
-                    dbg!(ctx.used_rect());
+                    unsafe {
+                        TOOLBAR_HEIGHT = ctx.used_rect().height();
+                    }
```
removed in [a8e1e5b](https://github.com/servo/servo/pull/29976/commits/a8e1e5be3ed5810dc0cbc0215d1d96b3e92ab612). thanks

atbrakhi committed:
winit: location field should reflect the current top-level url

delan committed:
merge from master

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

delan committed:
[NFC] winit: move Minibrowser out of App::run

delan committed:
winit: clean up toolbar height code

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

delan committed:
winit: make App own the Minibrowser if any

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

delan committed:
winit: make the go button work

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

atbrakhi committed:
winit:make the location field reflect the current top-level url

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

delan committed:
merge from master

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

atbrakhi committed:
winit: allow enabling minibrowser from command line

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

delan committed:
winit: tell compositor to repaint WR and flush when we repaint

delan committed:
winit: fix bug where location field edits would get overridden

delan committed:
winit: borrow the minibrowser once in App::handle_events

delan committed:
winit: address todo about viewport origin coordinates

delan committed:
winit: fix some minor problems with comments and errors

delan committed:
winit: update location field once per HistoryChanged event

delan committed:
winit: rename Window::set_toolbar_size to set_toolbar_height

delan committed:
winit: take toolbar height into account in hit testing

wusyong commented on ports/winit/app.rs:
```diff
@@ -114,7 +202,24 @@ impl App {
+            let response = match e {
+                winit::event::Event::WindowEvent { ref event, .. } => {
+                    if let Some(mut minibrowser) = app.minibrowser() {
+                        minibrowser.context.on_event(&event)
```
There are a few event surges that will return `repaint: true` and make the response of the browser pretty laggy.
These include `WindowEvent::CursorMoved` `WindowEvent::MouseWheel` and `WindowEvent::AxisMotion`.
Not sure what's the proper way to avoid, but I think one workaround could be detecting the height since this is the only egui part it has right now.

delan replied:
thanks! i noticed that too, but i wasn’t sure how to fix it yet ^^;;

delan replied:
fixed in 62ec187c2d8f89957f64c10a2cd1320ef47ef145 by using winit redraw requests to coalesce minibrowser updates.

ffc971de935aa90979213db8adc8472251e82ecc also tried filtering out CursorMoved events below the toolbar, but this made it impossible to unfocus the location bar by clicking the page contents, so i abandoned that.

delan committed:
winit: pass egui only relevant CursorMoved events

delan committed:
winit: scratch that, coalesce minibrowser updates instead

delan committed:
ensure both minibrowser and WR are repainted on every frame

delan committed:
compositing: only skip framebuffer clear in external present mode

delan committed:
winit: destroy egui glow Painter when shutting down

delan committed:
winit: clean up and fix license lint

delan committed:
fix duplicate versions lint by downgrading bytemuck_derive

was egui_glow ^0.22.0 (0.22.0)
→ egui/bytemuck ^0.22.0 (0.22.0)
→ epaint/bytemuck ^0.22.0 (0.22.0)
→ bytemuck ^1.7.2 (1.13.1)
→ bytemuck_derive ^1.4 (1.4.1)
→ syn ^2.0.1 (2.0.28)

now lock has bytemuck_derive 1.4.0
→ syn ^1.0.99 (1.0.103)

delan committed:
fix duplicate versions lint by disabling egui-winit/links

(we don’t need support for hyperlinks in our use of egui)

delan committed:
squelch duplicate versions lint by excluding clipboard-win

delan committed:
winit: fix compile warnings

delan committed:
winit: make gleam an optional dependency under /minibrowser

mrobinson commented on ports/winit/Cargo.toml:
```diff
@@ -30,12 +30,13 @@ ProductName = "Servo"
 [features]
 debugmozjs = ["libservo/debugmozjs"]
-default = ["webdriver", "max_log_level"]
+default = ["webdriver", "max_log_level", "minibrowser"]
```
I wonder if we could do away with this feature entirely. I think that the winit port doesn't use conditional compilation, so removing default features would break the build.

delan replied:
done in 00d61872126774efdf70f91a26074ae1332c1a2e

delan replied:
done in 00d61872126774efdf70f91a26074ae1332c1a2e

mrobinson commented on ports/winit/app.rs:
```diff
@@ -29,6 +34,62 @@ pub struct App {
+    minibrowser: Option<RefCell<Minibrowser>>,
+}
+
+struct Minibrowser {
```
Perhaps this could move to a separate file just for cleanliness? Maybe `minibrowser.rs`.

delan replied:
done in 06e0d2987e42a1fd542a34697be3f7ca37ecdc62

mrobinson commented on ports/winit/app.rs:
```diff
@@ -60,51 +121,107 @@ impl App {
+            minibrowser: None,
         };
 
+        if opts::get().minibrowser && window.winit_window().is_some() {
+            // Make sure the gl context is made current.
+            let webrender_surfman = window.webrender_surfman();
+            let webrender_gl = match webrender_surfman.connection().gl_api() {
+                GLApi::GL => unsafe { gl::GlFns::load_with(|s| webrender_surfman.get_proc_address(s)) },
+                GLApi::GLES => unsafe {
+                    gl::GlesFns::load_with(|s| webrender_surfman.get_proc_address(s))
+                },
+            };
+            let gl = unsafe {
+                glow::Context::from_loader_function(|s| {
+                    webrender_surfman.get_proc_address(s)
+                })
+            };
+            webrender_surfman.make_gl_context_current().unwrap();
+            debug_assert_eq!(webrender_gl.get_error(), gleam::gl::NO_ERROR,);
+
+            // Set up egui context for minibrowser ui
+            // Adapted from https://github.com/emilk/egui/blob/9478e50d012c5138551c38cbee16b07bc1fcf283/crates/egui_glow/examples/pure_glow.rs
+            app.minibrowser = Some(Minibrowser {
+                context: egui_glue::EguiGlow::new(events_loop.as_winit(), Arc::new(gl), None),
+                event_queue: RefCell::new(vec![]),
+                toolbar_height: 0f32.into(),
+                location: RefCell::new(ServoUrl::into_string(get_default_url())),
+                location_dirty: false.into(),
+            }.into());
+        }
```
This method is pretty long. Perhaps this could move into a helper?

mrobinson commented on ports/winit/app.rs:
```diff
@@ -60,51 +121,107 @@ impl App {
+                })
+            };
+            webrender_surfman.make_gl_context_current().unwrap();
+            debug_assert_eq!(webrender_gl.get_error(), gleam::gl::NO_ERROR,);
```
Looks like a trailing comma snuck in after `gleam::gl::NO_ERROR`.

delan replied:
fixed in ebdb00783024faa76abea961aced11a746b95ac4

delan replied:
fixed in ebdb00783024faa76abea961aced11a746b95ac4

mrobinson commented on ports/winit/browser.rs:
```diff
@@ -31,6 +31,8 @@ use tinyfiledialogs::{self, MessageBoxIcon, OkCancel, YesNo};
 
 pub struct Browser<Window: WindowPortsMethods + ?Sized> {
     current_url: Option<ServoUrl>,
+    current_url_taken: bool,
```
Hrm. I think maybe it would be a little simpler to just store a stringified version of the URL here when it changes. Then you could only do the conversion when the URL actually change and pass and expose a reference to a getter.

delan replied:
great idea, that makes a lot more sense! done in b01b10c591801be4736ccd501cfa292c9a361bc7

mrobinson commented on ports/winit/app.rs:
```diff
@@ -183,14 +337,45 @@ impl App {
+
+    /// Updates the location field when the [Browser] says it has changed, unless the user has
+    /// started editing it without clicking Go.
+    fn update_location_in_toolbar(&self, minibrowser: &mut RefMut<Minibrowser>) -> bool {
+        if !minibrowser.location_dirty.get() {
+            if let Some(current_url) = self.browser.borrow_mut().take_top_level_url_change() {
+                minibrowser.location = RefCell::new(ServoUrl::into_string(current_url));
+                return true;
+            }
+        }
```
If you take my suggestion below, I think this could be a function on the Minibrowser struct itself.

delan replied:
done in 06e0d2987e42a1fd542a34697be3f7ca37ecdc62

mrobinson commented on ports/winit/app.rs:
```diff
@@ -183,14 +337,45 @@ impl App {
 
+    /// Takes any outstanding events from the [Minibrowser], converting them to [EmbedderEvent] and
+    /// routing those to the App event queue.
+    fn queue_embedder_events_for_minibrowser_events(&self, minibrowser: &mut RefMut<Minibrowser>) {
+        for event in minibrowser.event_queue.borrow_mut().drain(..) {
+            match event {
+                MinibrowserEvent::Go => {
+                    let browser_id = self.browser.borrow().browser_id().unwrap();
+                    let location = minibrowser.location.borrow();
+                    let Ok(url) = ServoUrl::parse(&location) else {
+                        warn!("failed to parse location");
+                        break;
+                    };
+                    self.event_queue.borrow_mut().push(EmbedderEvent::LoadUrl(browser_id, url));
+                },
+            }
+        }
+    }
```
Could this be a function on the Minibrowser struct that takes a mutable reference to the event queue? That would further isolate the minibrowser code in its own code unit.

delan replied:
done in 06e0d2987e42a1fd542a34697be3f7ca37ecdc62

mrobinson commented on ports/winit/browser.rs:
```diff
@@ -437,6 +456,7 @@ where
                 },
                 EmbedderMsg::HistoryChanged(urls, current) => {
                     self.current_url = Some(urls[current].clone());
+                    self.current_url_taken = false;
```
Maybe just store the stringified version here.

delan replied:
done in b01b10c591801be4736ccd501cfa292c9a361bc7

mrobinson reviewed (changes requested):
Thanks for walking through this with us. I'm excited about this functionality! I just have a few suggestions.

delan committed:
winit: remove cargo feature, since it’s not really optional

delan committed:
winit: extract Minibrowser and related code to separate module

delan committed:
winit: remove unnecessary trailing comma

delan committed:
winit: simplify the ServoUrl serialisation optimisation

mrobinson reviewed (approved):
Thanks for these updates. This looks great!
